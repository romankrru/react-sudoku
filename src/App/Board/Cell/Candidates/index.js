import React, { memo } from "react";
import _ from "lodash";
import cn from "classnames";
import T from "prop-types";

import styles from "./index.module.css";

const Candidates = (props) => (
	<div className={styles.candidates}>
		{_.range(1, 10).map((n) => {
			return (
				<div
					onMouseDown={() => {
						if (props.isMobile || !props.isFocus) return;
						props.toggleCandidate(props.rowIdx, props.colIdx, n);
					}}
					key={n}
					className={cn(styles.candidateCell, {
						[styles.candidateCellActive]: props.candidates.includes(n),
						[styles.candidateCellHidden]:
							!props.candidates.includes(n) && !props.isFocus,
					})}
				>
					<span>{n}</span>
				</div>
			);
		})}
	</div>
);

Candidates.propTypes = {
	isFocus: T.bool.isRequired,
	toggleCandidate: T.func.isRequired,
	rowIdx: T.number.isRequired,
	colIdx: T.number.isRequired,
	isMobile: T.bool.isRequired,
	candidates: T.arrayOf(T.number).isRequired,
};

export default memo(Candidates);
