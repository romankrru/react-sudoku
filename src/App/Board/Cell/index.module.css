.cell {
    width: 50px;
    height: 50px;

    font-size: 36px;
    line-height: 50px;
    text-align: center;

    background: white;
    border-right: 1px solid #EDF1F8;
    border-bottom: 1px solid #EDF1F8;

    cursor: pointer;
}

.changeableCell {
    color: #2979FA;
}

.cellHighlighted {
    background-color: #F3F8FF;
}

.changeableCellInvalid {
    color: #FA2981;
}

.nativeCellInvalid {
    position: relative;
}

.nativeCellInvalid::before {
    content: "";

    position: absolute;
    top: 41px;
    left: 22px;

    width: 6px;
    height: 6px;

    border-radius: 50%;
    background-color: #FA2981;
}

.cellFocus {
    position: relative;
    background-color: white;
}

.cellFocus::after {
    content: "";
    pointer-events: none;

    position: absolute;
    top: 0px;
    left: 0px;

    width: 50px;
    height: 50px;
    
    border: 2px solid #2979FA;
    border-radius: 6px;
}

.cell:nth-child(n+19):nth-child(-n+27),
.cell:nth-child(n+46):nth-child(-n+54) {
    border-bottom: 1px solid #E7EFFF;
}

.cell:nth-child(n+28):nth-child(-n+36),
.cell:nth-child(n+55):nth-child(-n+64) {
    border-top: 1px solid #E7EFFF;
}

.cell:nth-child(3n) {
    border-right: 1px solid #E7EFFF;
}

.cell:nth-child(3n + 1) {
    border-left: 1px solid #E7EFFF;
}

.cell:nth-child(9n) {
    border-right: none;
}

.cell:nth-child(9n + 1) {
    border-left: none;
}

.cell:nth-child(n + 73) {
    border-bottom: none;
}