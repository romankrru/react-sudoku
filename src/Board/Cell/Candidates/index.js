import React from "react";
import cn from "classnames";
import styles from "./index.module.css";

const range = Array.from({ length: 9 }, (_, i) => i + 1);

const Candidates = (props) => {
	return (
		<div className={styles.candidates}>
			{range.map((n) => {
				return (
					<div
						{...(props.isFocus && {
							onClick: () =>
								props.toggleCandidate(props.rowIdx, props.colIdx, n),
						})}
						key={n}
						className={cn(styles.candidateCell, {
							[styles.candidateCellActive]: props.candidates.includes(n),
							[styles.candidateCellHidden]:
								!props.candidates.includes(n) && !props.isFocus,
						})}
					>
						{n}
					</div>
				);
			})}
		</div>
	);
};

export default Candidates;
