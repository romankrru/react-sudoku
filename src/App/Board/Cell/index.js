import React, { memo } from "react";
import cn from "classnames";
import T from "prop-types";

import Candidates from "./Candidates";
import styles from "./index.module.css";

const Cell = (props) => (
	<div className={styles.cellContainer}>
		<div
			onMouseDown={() => props.setFocus([props.rowIdx, props.colIdx])}
			className={cn(styles.cell, {
				[styles.cellFocus]: props.isFocus,
				[styles.cellHighlighted]: props.isHighlighted,
				[styles.changeableCell]: props.changeable,
				[styles.changeableCellInvalid]: !props.isValid && props.changeable,
				[styles.nativeCellInvalid]: !props.isValid && !props.changeable,
			})}
		>
			{props.value && <span>{props.value}</span>}

			{props.candidates && !props.value && (
				<Candidates
					rowIdx={props.rowIdx}
					colIdx={props.colIdx}
					isMobile={props.isMobile}
					toggleCandidate={props.toggleCandidate}
					candidates={props.candidates}
					isFocus={props.isFocus}
				/>
			)}
		</div>
	</div>
);

Cell.propTypes = {
	isMobile: T.bool.isRequired,
	setFocus: T.func.isRequired,
	rowIdx: T.number.isRequired,
	colIdx: T.number.isRequired,
	isFocus: T.bool.isRequired,
	isHighlighted: T.bool.isRequired,
	changeable: T.bool.isRequired,
	isValid: T.bool.isRequired,
	value: T.number,
	candidates: T.arrayOf(T.number).isRequired,
	toggleCandidate: T.func.isRequired,
};

export default memo(Cell);
