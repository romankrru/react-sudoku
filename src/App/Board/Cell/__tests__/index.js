import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Cell from "../";

const renderCell = () => {
	const defaultProps = {
		isMobile: false,
		setFocus: jest.fn,
		rowIdx: 0,
		colIdx: 0,
		isFocus: false,
		isHighlighted: false,
		changeable: true,
		isValid: true,
		candidates: [],
		toggleCandidate: jest.fn,
	};

	const view = render(<Cell {...defaultProps} />);

	const rerender = (newProps) =>
		view.rerender(<Cell {...{ ...defaultProps, ...newProps }} />);

	return { view, rerender };
};

describe("<Cell />", () => {
	it("should render correctly", () => {
		const { view } = renderCell();
		expect(view.container).toMatchSnapshot();
	});

	it("should set focus on mousedown", () => {
		const { view, rerender } = renderCell();

		const setFocusMock = jest.fn(() => {});

		rerender({
			setFocus: setFocusMock,
			colIdx: 2,
			rowIdx: 3,
		});

		expect(setFocusMock).not.toHaveBeenCalled();
		fireEvent.mouseDown(view.container.querySelector(".cell"));
		expect(setFocusMock).toHaveBeenCalled();
		expect(setFocusMock).toHaveBeenCalledWith([3, 2]);
	});
});
