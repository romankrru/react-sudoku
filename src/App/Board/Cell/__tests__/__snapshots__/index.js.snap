// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> should render correctly 1`] = `
<div>
  <div
    class="cellContainer"
  >
    <div
      class="cell changeableCell"
    >
      <div
        class="candidates"
      >
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            1
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            2
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            3
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            4
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            5
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            6
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            7
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            8
          </span>
        </div>
        <div
          class="candidateCell candidateCellHidden"
        >
          <span>
            9
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
